articles/clojure_elixir_fibonacci_mstalker/article.md

```diff
@@ -0,0 +1,339 @@
+
+
+### Clojure
+
+A straightforward solution to our problem looks like this in Clojure:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (= n 0) 0
+    (= n 1) 1
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We created a `fibonacci` namespace on the first line, and then we used `defn` to define a `compute` function. The next line, `[n]`, indicates that the function has one argument, which we call `n`.
+
+You'll notice all the parentheses. A pair of parentheses creates a list. The first element of a list is generally a function or macro, and the rest of the elements are the arguments.
+
+`cond` begins a conditional. If `n` is equal to zero, the function returns `0`. If `n` is `1`, then `compute` returns `1`. Mathematical operators like `=`, `+`, and `-` are just functions in Clojure, and they follow the same syntactical rules. That's why these operators appear before their arguments, rather than between them.
+
+You might have observed that no commas separate function arguments. That's normal in Clojure. Commas are optional in collections.
+
+Finally, we use the `:else` keyword to do something if `n` isn't `0` or `1`. We recursively call the `compute` function twice, passing in different arguments. This sums the two previous Fibonacci values. Putting the two `compute` function calls on separate lines is aesthetic. I could have written this:
+
+```clojure
+:else (+ (compute (- n 1)) (compute (- n 2)))
+```
+
+Clojure has no explicit `return`, so functions just return the last value they evaluate.
+
+If we wanted to, we could shorten the implementation slightly by using an inequality:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (< n 2) n
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We can call our function like this:
+
+```clojure
+(fibonacci/compute 10)
+;; 55
+```
+
+A semicolon starts a comment in Clojure. I wrote the return value of our function call after `;;`. Using two semicolons is idiomatic, and the language would have allowed us to use just one.
+
+
+### Elixir
+
+How does Elixir handle the simple implementation? Here is how we could rewrite the Clojure code in Elixir:
+
+```elixir
+defmodule Fibonacci do
+  def compute(n) do
+    cond do
+      n == 0 -> 0
+      n == 1 -> 1
+      true -> compute(n - 1) + compute(n - 2)
+    end
+  end
+end
+```
+
+First, I defined a module with `defmodule`. Modules are like Clojure's namespaces, and they contain zero or more functions. Then, I used `def` to define a function named `compute/1`. This function takes argument, `n`.
+
+Did you notice that I called this function `compute/1` instead of `compute`? We use both the function name and the number of arguments when referring to functions in Elixir.
```

Comment:
Does Elixir support [function overloading](https://en.wikipedia.org/wiki/Function_overloading)? Does Clojure?
Probably worth mentioning that here.

Reply:
Clojure supports function overloading. I'm not sure if what Elixir does is technically function overloading or not. I like the idea of exploring overloading, but I think I'm going to omit mentioning it here.